### PR TITLE
Parallelize sortAndDetermineBufferLayout

### PR DESCRIPTION
**Pull request by masterleinad**
In combination with the `CUDA-aware` MPI pull request (#162), we should also be able to avoid copying `permutation_indices` to the CPU.

### CODE REVIEW

**Comment by masterleinad** [comment]
This is what I tried first. At least in serial, it is way more expensive then what we are doing here.

**Comment by masterleinad** [comment]
It seems that the `OpenMP` backend still makes problems here.

**Comment by aprokop** [comment]
> This is what I tried first. At least in serial, it is way more expensive then what we are doing here.

Interesting. Hmm, lets think about this. I guess this situation is very different from the regular sorting in that here you would have few different values in a very large array. Regular sort algorithms would certainly struggle.

Is there any other algorithm that would work well in parallel that could do this in a single pass?

**Comment by aprokop** [comment]
And a completely orthogonal question: what do we know about the provided `ranks` that are coming from upstream? Could they already be sorted? For example, one of the paths goes like `communicateResultsBack` -> `createFromSends` -> `sortAndDetermineBufferLayout`. But the ranks that are being passed are setup as
```c++
  Kokkos::parallel_for(ARBORX_MARK_REGION("setup_communication_plan"),                                                                                       
                       Kokkos::RangePolicy<ExecutionSpace>(0, n_fwd_queries),                                                                                
                       KOKKOS_LAMBDA(int q) {                                                                                                                
                         for (int i = offset(q); i < offset(q + 1); ++i)                                                                                     
                         {                                                                                                                                   
                           export_ranks(i) = ranks(q);                                                                                                       
                         }                                                                                                                                   
                       });     
```
So, in fact, they are already chunked, so we could do a much easier routine. An additional implicit question here is whether we *need* them to be sorted in reverse manner, or if that does not matter.

**Comment by aprokop** [comment]
One other thing: Kokkos has `UnorderedMap`. [Here](https://github.com/kokkos/kokkos-tutorials/blob/master/Intro-Full/Exercises/unordered_map/Begin/unordered_map.cpp) is an example. I think it's worth trying it out if only for comparison.

File: src/details/ArborX_DetailsDistributor.hpp [hunk 1]
```diff
@@ -40,7 +81,7 @@ static void sortAndDetermineBufferLayout(InputView ranks,
   ARBORX_ASSERT(unique_ranks.empty());
   ARBORX_ASSERT(offsets.empty());
   ARBORX_ASSERT(counts.empty());
-  ARBORX_ASSERT(permutation_indices.extent(0) == ranks.extent(0));
+  ARBORX_ASSERT(permutation_indices.extent_int(0) == ranks.extent_int(0));
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply src/details/ArborX_DetailsDistributor.hpp]: Not necessary, adding consistency when we initialize `n` below and treating the size as `int` all the time below.

File: src/details/ArborX_DetailsDistributor.hpp [hunk 2]
```diff
@@ -55,31 +96,42 @@ static void sortAndDetermineBufferLayout(InputView ranks,
...
+  // case when (R) is small, but results may vary
+  using DeviceType = typename InputView::traits::device_type;
+  using ExecutionSpace = typename InputView::traits::execution_space;
 
+  Kokkos::View<int *, DeviceType> device_ranks_duplicate(
+      Kokkos::ViewAllocateWithoutInitializing(ranks.label()), ranks.size());
+  Kokkos::deep_copy(device_ranks_duplicate, ranks);
+  Kokkos::View<int *, DeviceType> device_permutation_indices(
+      Kokkos::ViewAllocateWithoutInitializing(permutation_indices.label()),
+      permutation_indices.size());
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply src/details/ArborX_DetailsDistributor.hpp]:
```
/scratch/kokkos-install-alt/include/Kokkos_View.hpp(2098): error: static assertion failed with "View allocation constructor requires managed memory"
```

File: src/details/ArborX_DetailsDistributor.hpp [hunk 3]
```diff
@@ -55,31 +96,42 @@ static void sortAndDetermineBufferLayout(InputView ranks,
...
-  Kokkos::deep_copy(ranks_duplicate, ranks);
+  // this implements a "sort" which is O(N * R) where (R) is the total number of
+  // unique destination ranks. it performs better than other algorithms in the
+  // case when (R) is small, but results may vary
+  using DeviceType = typename InputView::traits::device_type;
+  using ExecutionSpace = typename InputView::traits::execution_space;
 
+  Kokkos::View<int *, DeviceType> device_ranks_duplicate(
+      Kokkos::ViewAllocateWithoutInitializing(ranks.label()), ranks.size());
+  Kokkos::deep_copy(device_ranks_duplicate, ranks);
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply src/details/ArborX_DetailsDistributor.hpp]: No, `DeviceType` is `InputView::traits::device_type` so they have the same memory space (and typically a parallel execution space).
Still, I get
```
/scratch/kokkos-install-alt/include/Kokkos_View.hpp(2098): error: static assertion failed with "View allocation constructor requires managed memory"
```
I can investigate later if you care.

**Comment by masterleinad** [comment]
> One other thing: Kokkos has `UnorderedMap`. [Here](https://github.com/kokkos/kokkos-tutorials/blob/master/Intro-Full/Exercises/unordered_map/Begin/unordered_map.cpp) is an example. I think it's worth trying it out if only for comparison.

How would you want to use `UnorderedMap` here?

**Comment by aprokop** [comment]
> How would you want to use `UnorderedMap` her

Just like a regular hash table. For every element in the array, increment `map[array[i]]++`, then go through map to construct the output arrays.

**Comment by masterleinad** [comment]
Results for testing this with
- 60 MPI processes, 
- 10^7 points/MPI process,
- 10^6 queries/MPI process,
- and a varying number of neighbors for the knn search,
- and CUDA

using the `distributed_tree_driver` benchmark.

|  neighbors | new   |  old   |
|------------|-------|--------|
|  20         | 8.51e0 | 9.08e0 |
| 40          | 2.06e1 | 2.21e1 |
| 80          | 7.38e1 | 7.80e1 |
| 100        | 1.08e2 | 1.15e2 |
| 120        | 1.51e2 | 1.57e2 |
| 140        | 1.97e2 | 2.07e2 |

So for this setup, the new results are around 5% faster.

**Comment by dalg24** [comment]
image

**Comment by dalg24** [comment]
Just out of curiosity, did you look at a finer granularity on these timings?

**Comment by masterleinad** [comment]
No, `MAP` doesn't allow me to look into `sortAndDetermineBufferLayout`. We should be able to avoid a few additional `deep_copy` calls using this pull request since we can keep the permutation on the device.

Now, I discovered that again `sortResults` takes most of the time and I assume to get much more improvements by using `thrust` instead.

**Comment by masterleinad** [comment]
> Just out of curiosity, did you look at a finer granularity on these timings?

Looking at the profiling in #189 for the largest case (140 neighbors), we spend 16% computing the maximum, 78% in the parallel scan and 6% in copying the permutation indices back.

**Comment by dalg24** [comment]
retest this please

**Comment by masterleinad** [comment]
I don't get why the CI doesn't find MPI for PGI...

**Comment by aprokop** [comment]
> I don't get why the CI doesn't find MPI for PGI...

Because it's Wednesday :)

**Comment by masterleinad** [comment]
retest this please

File: src/details/ArborX_DetailsDistributor.hpp [hunk 4]
```diff
@@ -57,31 +58,50 @@ static void sortAndDetermineBufferLayout(InputView ranks,
...
-      if (ranks_duplicate(i) == largest_rank)
-      {
-        ranks_duplicate(i) = -1;
-        permutation_indices(i) = offsets.back() + counts.back();
-        ++counts.back();
-      }
-    }
-    offsets.push_back(offsets.back() + counts.back());
+    int result = 0;
+    Kokkos::parallel_scan("process_biggest_rank_items",
```
Comment (aprokop) [inline src/details/ArborX_DetailsDistributor.hpp]: `ARBORX_MARK_REGION`.

**Comment by masterleinad** [comment]
@dalg24 Here you go.

**Pull request merged**